File: backend/internal/httpserver/practice.go
package httpserver

import (
	"net/http"
	"slices"
	"strconv"
	"strings"

	"github.com/jackc/pgx/v5/pgxpool"

	"openwook/internal/api"
	"openwook/internal/auth"
	"openwook/internal/redisx"
	"openwook/internal/services"
)

type practiceStartRequest struct {
	BankID         *int64  `json:"bankId"`
	SessionType    *string `json:"sessionType"`
	QuestionCount  *int    `json:"questionCount"`
	Mode           *string `json:"mode"`
	QuestionTypeID *string `json:"questionTypeId"`
	AllQuestions   *bool   `json:"allQuestions"`
}

type practiceAnswerRequest struct {
	QuestionID    *int64         `json:"questionId"`
	AnswerPayload map[string]any `json:"answerPayload"`
	DurationMS    *int           `json:"durationMs"`
}

type offlinePracticeRequest struct {
	BankID  *int64 `json:"bankId"`
	Answers []struct {
		QuestionID    *int64         `json:"questionId"`
		AnswerPayload map[string]any `json:"answerPayload"`
	} `json:"answers"`
}

func BuildPracticeHandlers(pool *pgxpool.Pool, resolve auth.CurrentUserResolver) PracticeHandlers {
	return PracticeHandlers{
		List: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			user, err := auth.RequireUser(r, resolve)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			limit, err := queryPageLimit(r, 100)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			options := services.PracticeSessionListOptions{
				Status: strings.TrimSpace(r.URL.Query().Get("status")),
				Cursor: r.URL.Query().Get("cursor"),
				Limit:  limit,
			}
			data, err := services.ListPracticeSessions(r.Context(), pool, user, options)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			api.OK(w, r, data.Items, paginationMeta(data.PageInfo))
		}),
		Start: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			user, err := auth.RequireUser(r, resolve)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			body, err := decodeJSONBodyStrict[practiceStartRequest](r)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			input, err := parsePracticeStartInput(body)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			data, err := services.StartPracticeSession(r.Context(), pool, user, input)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			api.Created(w, r, data, nil)
		}),
		Get: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			user, err := auth.RequireUser(r, resolve)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			sessionID, err := parsePathID(r, "sessionId")
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			data, err := services.GetPracticeSession(r.Context(), pool, user, sessionID)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			api.OK(w, r, data, nil)
		}),
		QuestionPage: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			user, err := auth.RequireUser(r, resolve)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			sessionID, err := parsePathID(r, "sessionId")
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			var index *int
			if raw := strings.TrimSpace(r.URL.Query().Get("index")); raw != "" {
				if parsed, err := strconv.Atoi(raw); err == nil {
					index = &parsed
				}
			}
			data, err := services.GetPracticeQuestionPage(r.Context(), pool, user, sessionID, services.PracticeQuestionPageOptions{Index: index})
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			api.OK(w, r, data, nil)
		}),
		Questions: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			user, err := auth.RequireUser(r, resolve)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			sessionID, err := parsePathID(r, "sessionId")
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			data, err := services.GetPracticeQuestions(r.Context(), pool, user, sessionID)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			api.OK(w, r, data, nil)
		}),
		Results: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			user, err := auth.RequireUser(r, resolve)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			sessionID, err := parsePathID(r, "sessionId")
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			data, err := services.GetPracticeResults(r.Context(), pool, user, sessionID)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			api.OK(w, r, data, nil)
		}),
		SubmitAnswer: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			user, err := auth.RequireUser(r, resolve)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			sessionID, err := parsePathID(r, "sessionId")
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			body, err := decodeJSONBodyStrict[practiceAnswerRequest](r)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			input, err := parsePracticeAnswerInput(body)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			data, err := services.SubmitAnswer(r.Context(), pool, user, sessionID, input)
			if err != nil {
				api.HandleError(w, r, err)
				return
			}
			api.Created(w, r, data, nil)
		}),
		Complete: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			handlePracticeTerminalState(w, r, resolve, pool, "completed")
		}),
		Abandon: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			handlePracticeTerminalState(w, r, resolve, pool, "abandoned")
		}),
		OfflineUpload: buildOfflinePracticeHandler(pool, resolve),
	}
}

func buildOfflinePracticeHandler(pool *pgxpool.Pool, resolve auth.CurrentUserResolver) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		user, err := auth.RequireUser(r, resolve)
		if err != nil {
			api.HandleError(w, r, err)
			return
		}
		body, err := decodeJSONBodyStrict[offlinePracticeRequest](r)
		if err != nil {
			api.HandleError(w, r, err)
			return
		}
		if body.BankID == nil || *body.BankID < 1 || len(body.Answers) < 1 || len(body.Answers) > 500 {
			api.HandleError(w, r, api.ValidationError([]api.ValidationDetail{{Field: "answers", Message: "bankId and 1-500 answers are required"}}))
			return
		}
		questionIDs := make(map[int64]struct{}, len(body.Answers))
		for index, answer := range body.Answers {
			if answer.QuestionID == nil || *answer.QuestionID < 1 {
				api.HandleError(w, r, api.ValidationError([]api.ValidationDetail{{Field: "answers." + strconv.Itoa(index) + ".questionId", Message: "must be positive"}}))
				return
			}
			if _, duplicate := questionIDs[*answer.QuestionID]; duplicate {
				api.HandleError(w, r, api.ValidationError([]api.ValidationDetail{{Field: "answers." + strconv.Itoa(index) + ".questionId", Message: "must be unique"}}))
				return
			}
			questionIDs[*answer.QuestionID] = struct{}{}
		}
		idempotencyKey := strings.TrimSpace(r.Header.Get("Idempotency-Key"))
		if idempotencyKey == "" {
			api.HandleError(w, r, api.ValidationError([]api.ValidationDetail{{Field: "Idempotency-Key", Message: "is required"}}))
			return
		}
		rdb := redisx.Client()
		if rdb == nil {
			api.HandleError(w, r, api.NewError(http.StatusServiceUnavailable, "IDEMPOTENCY_UNAVAILABLE", "Offline practice upload is temporarily unavailable", nil))
			return
		}
		sessionKey := redisx.RedisKey("offline-practice", user.ID, idempotencyKey)
		sessionID, _ := rdb.Get(r.Context(), sessionKey).Int64()
		var session *services.PracticeSession
		if sessionID > 0 {
			session, err = services.GetPracticeSession(r.Context(), pool, user, sessionID)
		} else {
			session, err = services.StartPracticeSession(r.Context(), pool, user, services.PracticeSessionInput{
				BankID:        *body.BankID,
				SessionType:   "practice",
				QuestionCount: len(body.Answers),
				Mode:          "all",
			})
			if err == nil {
				sessionID = session.ID
				err = rdb.Set(r.Context(), sessionKey, sessionID, idempotencyTTL).Err()
			}
		}
		if err != nil {
			api.HandleError(w, r, err)
			return
		}
		for _, answer := range body.Answers {
			if _, err := services.SubmitAnswer(r.Context(), pool, user, sessionID, services.PracticeAnswerInput{
				QuestionID:    *answer.QuestionID,
				AnswerPayload: answer.AnswerPayload,
			}); err != nil {
				api.HandleError(w, r, err)
				return
			}
		}
		session, err = services.GetPracticeSession(r.Context(), pool, user, sessionID)
		if err == nil && session.Status == "active" {
			session, err = services.CompletePracticeSession(r.Context(), pool, user, sessionID, "completed")
		}
		if err != nil {
			api.HandleError(w, r, err)
			return
		}
		api.OK(w, r, session, nil)
	})
}

func handlePracticeTerminalState(w http.ResponseWriter, r *http.Request, resolve auth.CurrentUserResolver, pool *pgxpool.Pool, status string) {
	user, err := auth.RequireUser(r, resolve)
	if err != nil {
		api.HandleError(w, r, err)
		return
	}
	sessionID, err := parsePathID(r, "sessionId")
	if err != nil {
		api.HandleError(w, r, err)
		return
	}
	data, err := services.CompletePracticeSession(r.Context(), pool, user, sessionID, status)
	if err != nil {
		api.HandleError(w, r, err)
		return
	}
	api.OK(w, r, data, nil)
}

func parsePracticeStartInput(body practiceStartRequest) (services.PracticeSessionInput, error) {
	var details []api.ValidationDetail
	var input services.PracticeSessionInput

	if body.BankID == nil {
		details = append(details, api.ValidationDetail{Field: "bankId", Message: "is required"})
	} else if *body.BankID <= 0 {
		details = append(details, api.ValidationDetail{Field: "bankId", Message: "must be a positive integer"})
	} else {
		input.BankID = *body.BankID
	}
	input.SessionType = parseOptionalPracticeEnum(body.SessionType, "sessionType", []string{"practice", "review", "exam"}, &details)
	if body.QuestionCount != nil {
		if *body.QuestionCount <= 0 || *body.QuestionCount > 500 {
			details = append(details, api.ValidationDetail{Field: "questionCount", Message: "must be a positive integer no greater than 500"})
		} else {
			input.QuestionCount = *body.QuestionCount
		}
	}
	input.Mode = parseOptionalPracticeEnum(body.Mode, "mode", []string{"all", "wrong", "by_type", "exam"}, &details)
	if body.QuestionTypeID != nil {
		questionTypeID := strings.TrimSpace(*body.QuestionTypeID)
		if questionTypeID == "" || len(questionTypeID) > 64 {
			details = append(details, api.ValidationDetail{Field: "questionTypeId", Message: "must be a non-empty string"})
		} else {
			input.QuestionTypeID = &questionTypeID
		}
	}
	if body.AllQuestions != nil {
		input.AllQuestions = *body.AllQuestions
	}

	if len(details) > 0 {
		return services.PracticeSessionInput{}, api.ValidationError(details)
	}
	return input, nil
}

func parsePracticeAnswerInput(body practiceAnswerRequest) (services.PracticeAnswerInput, error) {
	var details []api.ValidationDetail
	input := services.PracticeAnswerInput{AnswerPayload: body.AnswerPayload, DurationMS: body.DurationMS}
	if body.QuestionID == nil {
		details = append(details, api.ValidationDetail{Field: "questionId", Message: "is required"})
	} else if *body.QuestionID <= 0 {
		details = append(details, api.ValidationDetail{Field: "questionId", Message: "must be a positive integer"})
	} else {
		input.QuestionID = *body.QuestionID
	}
	if body.AnswerPayload == nil {
		details = append(details, api.ValidationDetail{Field: "answerPayload", Message: "is required"})
	}
	if body.DurationMS != nil && *body.DurationMS < 0 {
		details = append(details, api.ValidationDetail{Field: "durationMs", Message: "must be a non-negative integer"})
	}

	if len(details) > 0 {
		return services.PracticeAnswerInput{}, api.ValidationError(details)
	}
	return input, nil
}

func parseOptionalPracticeEnum(value *string, field string, allowed []string, details *[]api.ValidationDetail) string {
	if value == nil {
		return ""
	}
	trimmed := strings.TrimSpace(*value)
	if slices.Contains(allowed, trimmed) {
		return trimmed
	}
	*details = append(*details, api.ValidationDetail{Field: field, Message: "has an invalid value"})
	return ""
}

File: server/routes/practice.py
"""Practice, analytics, and search routes."""

from typing import Annotated, Any, Literal, Self

from fastapi import APIRouter, Header, Query, Request
from pydantic import Field, StringConstraints, model_validator

from .. import envelope, redisx
from ..middleware import IDEMPOTENCY_TTL_SECONDS
from ..services import analytics as analytics_svc
from ..services import practice as practice_svc
from ..services import search as search_svc
from . import deps

router = APIRouter()

PositiveInt = Annotated[int, Field(gt=0)]
QuestionType = Annotated[
    str, StringConstraints(strip_whitespace=True, min_length=1, max_length=64)
]


class StartSessionBody(deps.RequestBody):
    bank_id: PositiveInt
    session_type: Literal['', 'practice', 'review', 'exam'] = ''
    question_count: Annotated[int, Field(gt=0, le=500)] | None = None
    mode: Literal['', 'all', 'wrong', 'by_type', 'exam'] = ''
    question_type_id: QuestionType | None = None
    all_questions: bool = False


class SubmitAnswerBody(deps.RequestBody):
    question_id: PositiveInt
    answer_payload: Any
    duration_ms: Annotated[int, Field(ge=0)] | None = None


class OfflineAnswerBody(deps.RequestBody):
    question_id: PositiveInt
    answer_payload: dict[str, Any] = Field(default_factory=dict)


class OfflinePracticeBody(deps.RequestBody):
    bank_id: PositiveInt
    answers: list[OfflineAnswerBody] = Field(min_length=1, max_length=500)

    @model_validator(mode='after')
    def unique_questions(self) -> Self:
        question_ids = [answer.question_id for answer in self.answers]
        if len(question_ids) != len(set(question_ids)):
            raise ValueError('questionId values must be unique')
        return self


@router.get('/api/v1/practice-sessions')
async def list_sessions(
    request: Request,
    status: str = '',
    cursor: str = '',
    limit: deps.PageLimit100 = None,
    updated_since: deps.UpdatedSince = None,
):
    user = await deps.current_user(request)
    data = await practice_svc.list_practice_sessions(
        deps.pool(request),
        user,
        limit or 0,
        status,
        cursor,
        updated_since.isoformat() if updated_since else '',
    )
    return envelope.ok(request, data.items, data.page_info.as_meta())


@router.post('/api/v1/practice-sessions')
async def start_session(request: Request, body: StartSessionBody):
    user = await deps.current_user(request)
    data = await practice_svc.start_practice_session(
        deps.pool(request),
        user,
        practice_svc.PracticeSessionInput(
            bank_id=body.bank_id,
            session_type=body.session_type,
            question_count=body.question_count or 0,
            mode=body.mode,
            question_type_id=body.question_type_id,
            all_questions=body.all_questions,
        ),
    )
    return envelope.created(request, data)


@router.get('/api/v1/practice-sessions/{session_id}')
async def get_session(request: Request, session_id: deps.PositiveId):
    user = await deps.current_user(request)
    return envelope.ok(
        request,
        await practice_svc.get_practice_session(
            deps.pool(request), user, session_id
        ),
    )


@router.get('/api/v1/practice-sessions/{session_id}/question-page')
async def question_page(
    request: Request,
    session_id: deps.PositiveId,
    index: Annotated[int | None, Query(ge=0)] = None,
):
    user = await deps.current_user(request)
    data = await practice_svc.get_practice_question_page(
        deps.pool(request), user, session_id, index
    )
    return envelope.ok(request, data)


@router.get('/api/v1/practice-sessions/{session_id}/questions')
async def session_questions(request: Request, session_id: deps.PositiveId):
    user = await deps.current_user(request)
    return envelope.ok(
        request,
        await practice_svc.get_practice_questions(
            deps.pool(request), user, session_id
        ),
    )


@router.get('/api/v1/practice-sessions/{session_id}/results')
async def session_results(request: Request, session_id: deps.PositiveId):
    user = await deps.current_user(request)
    return envelope.ok(
        request,
        await practice_svc.get_practice_results(
            deps.pool(request), user, session_id
        ),
    )


@router.post(
    '/api/v1/practice-sessions/{session_id}/answers',
)
async def submit_answer(
    request: Request, session_id: deps.PositiveId, body: SubmitAnswerBody
):
    user = await deps.current_user(request)
    data = await practice_svc.submit_answer(
        deps.pool(request),
        user,
        session_id,
        practice_svc.PracticeAnswerInput(
            question_id=body.question_id,
            answer_payload=body.answer_payload
            if isinstance(body.answer_payload, dict)
            else {},
            duration_ms=body.duration_ms,
        ),
    )
    return envelope.created(request, data)


@router.post('/api/v1/practice-sessions/{session_id}/complete')
async def complete_session(request: Request, session_id: deps.PositiveId):
    user = await deps.current_user(request)
    return envelope.ok(
        request,
        await practice_svc.complete_practice_session(
            deps.pool(request), user, session_id, 'completed'
        ),
    )


@router.post('/api/v1/practice-sessions/{session_id}/abandon')
async def abandon_session(request: Request, session_id: deps.PositiveId):
    user = await deps.current_user(request)
    return envelope.ok(
        request,
        await practice_svc.complete_practice_session(
            deps.pool(request), user, session_id, 'abandoned'
        ),
    )


@router.post('/api/v1/offline-practice')
async def offline_practice(
    request: Request,
    body: OfflinePracticeBody,
    idempotency_key: Annotated[str, Header(alias='Idempotency-Key', min_length=1)],
):
    user = await deps.current_user(request)
    rdb = redisx.client()
    if rdb is None:
        raise envelope.new_error(
            503,
            'IDEMPOTENCY_UNAVAILABLE',
            'Offline practice upload is temporarily unavailable',
        )
    session_key = redisx.redis_key(
        'offline-practice', user.id, idempotency_key.strip()
    )
    pool = deps.pool(request)
    try:
        raw = await rdb.get(session_key)
        session_id = int(raw) if raw else 0
    except Exception:
        session_id = 0
    if session_id > 0:
        session = await practice_svc.get_practice_session(pool, user, session_id)
    else:
        session = await practice_svc.start_practice_session(
            pool,
            user,
            practice_svc.PracticeSessionInput(
                bank_id=body.bank_id,
                session_type='practice',
                question_count=len(body.answers),
                mode='all',
            ),
        )
        session_id = session['id']
        try:
            await rdb.set(
                session_key, session_id, ex=IDEMPOTENCY_TTL_SECONDS
            )
        except Exception:
            pass
    for answer in body.answers:
        await practice_svc.submit_answer(
            pool,
            user,
            session_id,
            practice_svc.PracticeAnswerInput(
                question_id=answer.question_id,
                answer_payload=answer.answer_payload,
            ),
        )
    session = await practice_svc.get_practice_session(pool, user, session_id)
    if session['status'] == 'active':
        session = await practice_svc.complete_practice_session(
            pool, user, session_id, 'completed'
        )
    return envelope.ok(request, session)


@router.get('/api/v1/analytics/me/summary')
async def analytics_summary(request: Request):
    user = await deps.current_user(request)
    return envelope.ok(
        request, await analytics_svc.get_analytics_summary(deps.pool(request), user)
    )


@router.get('/api/v1/analytics/me/snapshot')
async def analytics_snapshot(request: Request):
    user = await deps.current_user(request)
    return envelope.ok(
        request, await analytics_svc.get_user_stats_snapshot(deps.pool(request), user)
    )


@router.get('/api/v1/analytics/banks/{bank_id}')
async def analytics_bank(request: Request, bank_id: deps.PositiveId):
    user = await deps.current_user(request)
    return envelope.ok(
        request,
        await analytics_svc.get_bank_analytics(
            deps.pool(request), user, bank_id
        ),
    )


@router.get('/api/v1/analytics/banks/{bank_id}/leaderboard')
async def analytics_leaderboard(
    request: Request,
    bank_id: deps.PositiveId,
    limit: Annotated[int, Query(gt=0, le=100)] = 20,
):
    user = await deps.current_user(request)
    return envelope.ok(
        request,
        await analytics_svc.get_bank_leaderboard(
            deps.pool(request), user, bank_id, limit
        ),
    )


@router.get('/api/v1/analytics/imports/{job_id}')
async def analytics_import(request: Request, job_id: deps.PositiveId):
    user = await deps.current_user(request)
    return envelope.ok(
        request,
        await analytics_svc.get_import_analytics(
            deps.pool(request), user, job_id
        ),
    )


@router.get('/api/v1/search/{kind}')
async def search_kind(
    request: Request,
    kind: Literal['questions'],
    bank_id: Annotated[int | None, Query(alias='bankId', gt=0)] = None,
    type: str = '',
    status: str = '',
    limit: deps.PageLimit100 = None,
    cursor: str = '',
    q: str = '',
):
    user = await deps.current_user(request)
    async with deps.pool(request).connection() as conn:
        data = await search_svc.search_questions(
            conn,
            user,
            bank_id_raw=str(bank_id or ''),
            type_raw=type,
            status_raw=status,
            limit_raw=str(limit or ''),
            cursor_str=cursor,
            query=q,
        )
    return envelope.ok(request, data.items, data.page_info.as_meta())
